dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,142 @@
+import json
+import time
+import sys
+import curses
+from matplotlib import pyplot
+import kafka
+
+
+def populate(variables, decode,
+             x_axis_data, prepend=""):
+    """Takes the data from the dict and populates the variables"""
+    for other_key in decode:
+        valid = True
+        name = prepend+other_key
+        try:
+            following = float(decode[other_key])
+        except (ValueError, TypeError):
+            valid = False
+        if isinstance(decode[other_key], dict):
+            populate(variables,
+                     decode[other_key],
+                     x_axis_data,
+                     name+": ")
+        else:
+            if name in variables:
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(len(x_axis_data)-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+
+
+
+def check():
+    """Checks that the right number of args are present"""
+    topic = "default"
+    if (len(sys.argv)) < 2:
+        print("Please provide a topic name on the command line!")
+        print("Using default topic name 'default'...")
+    else:
+        topic = sys.argv[1]
+    return topic
+
+
+def redraw(figures, key):
+    """Redraw the given figure from a dict of figures"""
+    figures[key].canvas.draw()
+    figures[key].canvas.flush_events()
+
+
+
+
+
+def main():
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = curses.initscr()
+    curses.cbreak()
+    screen.nodelay(True)
+
+    topic = check()
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    x_axis_data = []
+    count = 0
+    pyplot.ion() #Interactive mode on
+    # To consume latest messages and auto-commit offsets
+    consumer = kafka.KafkaConsumer(topic,
+                                   group_id='my-group',
+                                   bootstrap_servers=['localhost:9092'],
+                                   auto_offset_reset='earliest',
+                                   enable_auto_commit=False)
+
+    exit_char = curses.ERR
+
+    for message in consumer:
+        try:
+            update = json.loads(message.value)
+        except json.decoder.JSONDecodeError as problem:
+            print("Error: Flawed JSON in value read from kafka topic: ",
+                  message.value)
+            print("Error message: ", problem)
+            update = ""
+        data.update(update)
+        print("%s:%d:%d: key=%s value=%s" % (message.topic, message.partition,
+                                             message.offset, message.key,
+                                             data))
```

Comment:
Consider using logging.

Reply:
The purpose of this consumer is to graph arbitrary json data - shouldn't kafka be logging everything that's written to it in any event?

Regardless, there's no particular reason to print this data, either. So the print statement has been taken out.